src/RamTransforms.cpp

```diff
@@ -564,17 +944,24 @@ std::unique_ptr<RamOperation> ChoiceConversionTransformer::rewriteIndexScan(cons
 
     // Convert the IndexScan/If pair into an IndexChoice
     if (transformTuple) {
-        std::vector<std::unique_ptr<RamExpression>> newValues;
+        RamPattern newValues;
         const auto* filter = dynamic_cast<const RamFilter*>(&indexScan->getOperation());
         const int identifier = indexScan->getTupleId();
         const RamRelation& rel = indexScan->getRelation();
 
-        for (auto& cur : indexScan->getRangePattern()) {
+        for (auto& cur : indexScan->getRangePattern().first) {
```

Comment:
`souffle::clone` is a family of overloads. They handle nullptrs (branch can be removed) and support the common case of `vector`s (including recursive containers). e.g. simplify to `queryPattern.first = clone(indexScan->getRangePattern().first());` If you add an overload for `std::pair` you could further simplify it to `clone(indexScan->getRangePattern());` and inline that into `mk<RamIndexChoice>(...)`.

Reply:
Util.h:507:50: error: ‘auto’ parameter not permitted in this context
  507 | std::vector<std::unique_ptr<A>> clone(const std::vector<<A*>>& xs) {
      |                                                  ^~~~~~

This is the error that I get when I try to create an overload for a vector of pointers. Keep in mind that this overload is required because getRangePattern() returns a vector of raw pointers (non-owning pointers) to the RamExpression for upper/lower bound vectors respectively.

At this point, I've just kept with cloning each element in a loop. Would be happy to change this is someone can help write an overload for this.

Reply:
Found a way of doing this. Dumb error of putting two sets of angle brackets when I only needed one...